pkg/ebpf/uprobes/attacher.go

```diff
@@ -0,0 +1,901 @@
+	Unregister(uint32) error
+	Clear()
+	GetRegisteredProcesses() map[uint32]struct{}
+}
+
+// AttachCallback is a callback that is called whenever a probe is attached successfully
+type AttachCallback func(*manager.Probe, *utils.FilePath)
+
+// UprobeAttacher is a struct that handles the attachment of uprobes to processes and libraries
+type UprobeAttacher struct {
+	name         string
+	done         chan struct{}
+	wg           sync.WaitGroup
+	config       AttacherConfig // Not a pointer, we want a copy of the configuration so that the user cannot change it, as we have certain cached values that we have no way to invalidate if the config changes after the attacher is created
+	fileRegistry FileRegistry
+	manager      ProbeManager
+	inspector    BinaryInspector
+
+	// pathToAttachedProbes maps a filesystem path to the probes attached to it. Used to detach them
+	// once the path is no longer used.
+	pathToAttachedProbes     map[string][]manager.ProbeIdentificationPair
+	onAttachCallback         AttachCallback
+	soWatcher                *sharedlibraries.EbpfProgram
+	handlesLibrariesCached   *bool
+	handlesExecutablesCached *bool
+}
+
+// NewUprobeAttacher creates a new UprobeAttacher. Receives as arguments the
+// name of the attacher, the configuration, the probe manager (ebpf.Manager
+// usually), a callback to be called whenever a probe is attached (optional, can
+// be nil), and the binary inspector to be used (e.g., while we usually want
+// NativeBinaryInspector here, we might want the GoBinaryInspector to attach to
+// Go functions in a different way).
+// Note that the config is copied, not referenced. The attacher caches some values
+// that depend on the configuration, so any changes to the configuration after the
+// attacher would make those caches incoherent. This way we ensure that the attacher
+// is always consistent with the configuration it was created with.
+func NewUprobeAttacher(name string, config AttacherConfig, mgr ProbeManager, onAttachCallback AttachCallback, inspector BinaryInspector) (*UprobeAttacher, error) {
+	config.SetDefaults()
+
+	if err := config.Validate(); err != nil {
+		return nil, fmt.Errorf("invalid attacher configuration: %w", err)
+	}
+
+	ua := &UprobeAttacher{
+		name:                 name,
+		config:               config,
+		fileRegistry:         utils.NewFileRegistry(name),
+		manager:              mgr,
+		onAttachCallback:     onAttachCallback,
+		pathToAttachedProbes: make(map[string][]manager.ProbeIdentificationPair),
+		done:                 make(chan struct{}),
+		inspector:            inspector,
+	}
+
+	utils.AddAttacher(name, ua)
+
+	return ua, nil
+}
+
+// SetRegistry allows changing the file registry used by the attacher. This is useful for testing purposes, to
+// replace the registry with a mock object
+func (ua *UprobeAttacher) SetRegistry(registry FileRegistry) {
+	ua.fileRegistry = registry
+}
```

Comment:
maybe let's move it to a testutils file (with a build flag of `test`) to prevent regular code to use it